commonmark/src/Commonmark/Nodes.hs

```diff
@@ -0,0 +1,208 @@
+{-# LANGUAGE ExistentialQuantification  #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE RecordWildCards            #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE UndecidableInstances       #-}
+module Commonmark.Nodes
+  ( Nodes (..)
+  , singleNode
+  , fromNodes
+  , Node (..)
+  , getNodeType
+  , SomeNodeType (..)
+  , NodeType (..)
+  , NodeTypeBlock (..)
+  , NodeTypeInline (..)
+  , -- * Helpers
+    mapNodes
+  , traverseNodes
+  , concatMapNodes
+  )
+where
+
+import           Commonmark.Types
+import           Data.Kind (Constraint)
+import           Data.Text (Text)
+import qualified Data.Text as T
+import           Data.Typeable        (Typeable, cast)
+
+-- | Nodes parsed from a markdown document, that can later be rendered
+-- to a value of type @a@ with 'fromNodes'.
+--
+-- An example filtering out all raw HTML content:
+--
+-- > let isRawHTML node
+-- >       | Just NodeRawInline{} <- getNodeType node = True
+-- >       | Just NodeRawBlock{} <- getNodeType node = True
+-- >       | otherwise = False
+-- > in Nodes . filter (not . isRawHTML) . unNodes
+newtype Nodes a = Nodes { unNodes :: [Node a] }
```

Comment:
These types have an annoying type parameter because in order to implement `fromNodes :: Nodes -> a`, you would need to tell the compiler that `a` can render any `NodeType`s that might be in the `Nodes`. But since `NodeType` is extensible, it might be possible `Nodes` to parse some new `FancyNode` and `Html ()` not know how to render it (for example).

A clearer type would be to say
```hs
data Nodes nodeTypes = ...
```
where `nodeTypes` is a type-level set of node types that are represented in a `Nodes` value, and `fromNodes` could have a constraint that says "`a` can render every one of these node types". But this isn't very feasible right now.

The current implementation effectively bundles the Dict for `IsBlock (Html ()) (Html ())` when parsing `SomeNodeType`, so each `Node` carries around how to "finish" rendering